client/lib/track-element-size/index.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * External dependencies
+ */
+import { useRef, useEffect, useState, useCallback } from 'react';
+import { throttle } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import afterLayoutFlush from 'lib/after-layout-flush';
+
+const THROTTLE_RATE = 200;
+
+/**
+ * React hook that subscribes a consumer to changes to the bounding client rect of an element, based
+ * on window resize events.
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @param {Function} callback The function to call back on changes. Takes a single parameter: `boundingClientRect`.
+ *
+ * @returns {Function} The ref to be set on the consumer component.
+ */
+export function useWindowResizeCallback( callback ) {
+	const lastRect = useRef( {} );
+	const [ element, setElement ] = useState( null );
+
+	const callbackRef = useCallback( node => {
+		if ( node ) {
+			setElement( node );
+		}
+	}, [] );
+
+	useEffect(() => {
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = rect => {
+			if ( rect ) {
+				const jsonRect = rect.toJSON ? rect.toJSON() : rect;
```

Comment:
TIL https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect#Notes

Browsers. Sigh.

I assume this toJSON is to convert the DOMRect to a plain object?

Reply:
That's right. It seemed like a saner API to return a JSON object that we assume can have different properties, rather than a `ClientRect` or `DOMRect`, depending on browser implementation.